MINOR: Add log segment unit tests, If the maximum offset beyond index, appen…

lanshiqin opened the pull request:
Add log segment unit tests, If the maximum offset beyond index, appended to the log section, it throws LogSegmentOffsetOverflowException

*More detailed description of change
The unit tests for the log segment should cover more scenario validation, and the unit tests submitted this time cover LogSegmentOffsetOverflowException that are expected to be thrown when the maximum index is exceeded

lanshiqin committed:
Add log segment unit tests, If the maximum offset beyond index, appended to the log section, it throws LogSegmentOffsetOverflowException

divijvaidya commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +66,21 @@ class LogSegmentTest {
+  def testAppendForLogSegmentOffsetOverflowException(): Unit = {
+    val seg = createSegment(0)
+    val largestOffset: Long = Integer.MAX_VALUE + 1
+    val largestTimestamp: Long = Time.SYSTEM.milliseconds()
```
In the spirit of keeping this test run fast and deterministic, could we hard code a value here?

lanshiqin replied:
Thanks, it really shouldn't be hardcoded directly here, I should use @ParameterizedTest

divijvaidya commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +66,21 @@ class LogSegmentTest {
+  @Test
+  def testAppendForLogSegmentOffsetOverflowException(): Unit = {
+    val seg = createSegment(0)
+    val largestOffset: Long = Integer.MAX_VALUE + 1
```
I would suggest to use `@ParameterizedTest` and also test for:
1. Negative value for largestOffset
2. Value exceeding the range of integer (existing)

lanshiqin replied:
Good idea

divijvaidya commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +66,21 @@ class LogSegmentTest {
+   */
+  @Test
+  def testAppendForLogSegmentOffsetOverflowException(): Unit = {
+    val seg = createSegment(0)
```
May I suggest to create a segment with non-zero index so that we can test for https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/AbstractIndex.java#L554 condition to be false. The `relativeOffset` should fit within an integer but the current test case only tests for situation where we have 0 as starting index.

lanshiqin replied:
Thanks, good advice, I've modified it along with this scheme

divijvaidya reviewed (changes requested):
Thank you for your contribution! I left some comments so that we can test multiple cases with this same unit test.

Hangleton commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +66,21 @@ class LogSegmentTest {
+    val shallowOffsetOfMaxTimestamp: Long = 0
+    val memoryRecords: MemoryRecords = records(0, "hello")
+    assertThrows(classOf[LogSegmentOffsetOverflowException], () => {
+      seg.append(largestOffset, largestTimestamp, shallowOffsetOfMaxTimestamp, memoryRecords)
```
nit: `largestTimestamp` could be misleading as it could make believe the timestamp itself is set to the maximum possible value.

lanshiqin replied:
Thanks, I should have changed the variable name to the 'currentTime'

Commit:
use @ParameterizedTest more test scenarios are covered.
1. Negative value for largestOffset
2. Value exceeding the range of integer
3.create a segment with non-zero index

lanshiqin commented:
Use @ParameterizedTest more test scenarios are covered.
1. Negative value for largestOffset
2. Value exceeding the range of integer 
3. Create a segment with non-zero index so that we can test for https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/AbstractIndex.java#L554 condition to be false.

Commit:
use @ParameterizedTest more test scenarios are covered.
1. Negative value for largestOffset
2. Value exceeding the range of integer
3.create a segment with non-zero index

divijvaidya commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +68,28 @@ class LogSegmentTest {
+   * If the maximum offset beyond index, appended to the log section, it throws LogSegmentOffsetOverflowException
+   */
+  @ParameterizedTest
+  @CsvSource(Array(
```
Note that baseOffset can be `Long` but `largestOffset - baseOffset` should be <= `Integer.MaxValue`. This case is missing from our test here.

Could we add the following test cases as well:

baseOffset is a number > Integer.MaxValue, largestOffset is a number > Integer.MaxValue, such that `largestOffset - baseOffset` > Integer.MaxValue. This will throw an error.

lanshiqin replied:
Thanks, I've added this use case data

divijvaidya reviewed (changes requested):
Thank you for making the suggested changes. 

One last suggestion to add a missing case.

lanshiqin committed:
Add test data coverage.
baseOffset is a number > Integer.MaxValue, largestOffset is a number > Integer.MaxValue, such that largestOffset - baseOffset > Integer.MaxValue. This will throw an error.

divijvaidya reviewed (approved):
Looks good to me! (and welcome to the Kafka community)

The failing tests are unrelated to this test case. As a next step, a committer will look into this and merge it. Perhaps @showuon will be interested.

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

showuon commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +68,29 @@ class LogSegmentTest {
   }
 
+  /**
+   * If the maximum offset beyond index, appended to the log section, it throws LogSegmentOffsetOverflowException
```
This comment is not clear enough (ex: what is maximum offset mean in this test? what index mean here?). Maybe:
```
LogSegmentOffsetOverflowException should be thrown while appending the logs if:
1. largestOffset - baseOffset < 0
2. largestOffset - baseOffset > Integer.MAX_VALUE
```

lanshiqin replied:
Good idea. I've modified it

showuon reviewed (commented):
LGTM, just a minor comment.

lanshiqin committed:
Modify a clearer description

showuon reviewed (approved):
LGTM! Will merge it after CI build completed. Thanks for the PR!

Hangleton commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +68,32 @@ class LogSegmentTest {
 
+  /**
+   * LogSegmentOffsetOverflowException should be thrown while appending the logs if:
+   * 1. largestOffset < 0
```
Nit: we can remove (1) since (1) => (2).

lanshiqin replied:
Thanks for your advice, I have revised the description.

Hangleton commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +68,32 @@ class LogSegmentTest {
+    "-2147483648, 0",
+    "2147483648,4294967296"
+  ))
+  def testAppendForLogSegmentOffsetOverflowException(baseOffset: Long, largestOffset: Long): Unit = {
```
What is the expected behaviour for `baseOffset == largestOffset`?

lanshiqin replied:
When baseOffset == largestOffset, the teardown method in the test class is called at the end of the test, which in turn calls the logSegment _.close() => timeIndex.maybeAppend, Eventually an exception will be thrown (Integer overflow for offset: 0), but this exception is not the point that this test method needs to cover

Hangleton replied:
I see, thanks for checking. Could we maybe confirm if the overflow exception comes from this line in `TimeIndex#maybeAppend`:

```
mmap.putInt(relativeOffset(offset))
```

And if that is the case, why is the relative offset < 0 or > max_int?

Hangleton replied:
I see, thanks for checking. Could we maybe confirm if the overflow exception comes from this line in `TimeIndex#maybeAppend`:

```
mmap.putInt(relativeOffset(offset))
```

And if that is the case, why is the relative offset < 0 or > max_int?

lanshiqin replied:
Sorry, I need to correct that. 
When baseOffset == largestOffset, LogSegment.append executes normally without throwing an exception. Because my unit test of parameters shallowOffsetOfMaxTimestamp value is 0, which can lead to rear unit test processing, TimeIndex# maybeAppend: In a `relativeOffset(offset),` the value of offset is 0. If baseOffset is a value greater than 0, the result will be <0, resulting in an exception.
I have repaired the shallowOffsetOfMaxTimestamp value of this test.

Hangleton replied:
Got it, thanks for confirming!

Hangleton commented on core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:
```diff
@@ -65,6 +68,32 @@ class LogSegmentTest {
+    "100, 10",
+    "2147483648, 0",
+    "-2147483648, 0",
+    "2147483648,4294967296"
```
nit: Note that the max int value on the Java platform is `2147483647` so we are exercising max int + 1.

lanshiqin replied:
Yes, it was intentionally written as 2147483648. The test method and LogSegment receive offset arguments of type Long in order to test scenarios exceeding Int.MAX_VALUE

Hangleton reviewed (commented):
Thanks for the PR. A few minor comments.

lanshiqin committed:
Modify a clearer description

Hangleton reviewed (commented):
Thanks @lanshiqin for your update. Added a follow-up comment, minor but I am curious since the case you mention involves an offset overflow as well.

lanshiqin committed:
ShallowOffsetOfMaxTimestamp values should be and largestOffset value consistent

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

lanshiqin committed:
Merge branch 'apache:trunk' into trunk

showuon commented:
@Hangleton , do you have any other comments? If no, I'm going to merge it this week.

Hangleton commented:
@showuon Thanks for the follow-up and apologies for the delay, the PR seems good to me.
